cupy/linalg/norms.py

```diff
@@ -69,36 +69,24 @@ def norm(x, ord=None, axis=None, keepdims=False):
         elif ord == 0:
             # Zero norm
             # Convert to Python float in accordance with NumPy
-            return (x != 0).sum(axis=axis, keepdims=keepdims, dtype='d')
+            return (x != 0).astype(x.real.dtype).sum(axis=axis, keepdims=keepdims)
         elif ord == 1:
             # special case for speedup
             return abs(x).sum(axis=axis, keepdims=keepdims)
         elif ord is None or ord == 2:
             # special case for speedup
-            if issubclass(x.dtype.type, numpy.complexfloating):
-                s = abs(x)
-                s *= s
-            else:
-                s = x ** 2
+            s = (x.conj() * x).real
             return cupy.sqrt(s.sum(axis=axis, keepdims=keepdims))
         else:
             try:
                 float(ord)
             except TypeError:
                 raise ValueError("Invalid norm order for vectors.")
 
-            # Mirror Numpy behavior of casting to double for non-complex
-            # dtypes, and to float32 or float64 for complex dtypes and
-            # no reduction over all axes.
-            cast_dtype = 'd'
-            if issubclass(x.dtype.type, numpy.complexfloating):
-                if keepdims or tuple(sorted(axis)) != tuple(range(nd)):
-                    cast_dtype = x.dtype.char.lower()  # 'D'->'d' and 'F'->'f'
-
-            absx = abs(x).astype(cast_dtype)
+            absx = abs(x)
             absx **= ord
             ret = absx.sum(axis=axis, keepdims=keepdims)
-            ret **= (1.0 / ord)
+            ret **= cupy.reciprocal(ord, dtype=ret.dtype)
```

Comment:
I think you need to update this - `np.reciprocal` fails if `ord` is an integer, so yours might too

Reply:
Thank you for the comment!
I guess both NumPy and CuPy work correctly if dtype is given to `reciprocal`?

```
>>> import numpy as np, cupy as cp
>>> np.reciprocal(10, dtype=np.float32)
0.1
>>> cp.reciprocal(10, dtype=cp.float32)
array(0.1, dtype=float32)
```

Reply:
I raised a PR regarding this, could you please confirm? https://github.com/numpy/numpy/pull/10667